eclair-core/src/main/scala/fr/acinq/eclair/channel/ChannelTypes.scala

```diff
@@ -271,15 +275,28 @@ case class ChannelVersion(bits: BitVector) {
 
 object ChannelVersion {
   import scodec.bits._
+
   val LENGTH_BITS: Int = 4 * 8
 
   private val USE_PUBKEY_KEYPATH_BIT = 0 // bit numbers start at 0
   private val USE_STATIC_REMOTEKEY_BIT = 1
+  private val USE_ANCHOR_OUTPUTS_BIT = 2
 
   private def setBit(bit: Int) = ChannelVersion(BitVector.low(LENGTH_BITS).set(bit).reverse)
 
+  def pickChannelVersion(localFeatures: Features, remoteFeatures: Features): ChannelVersion = {
```

Comment:
What do you mean by "combine" here?
I think we still want to allow channels to use `static_remotekey` without using `anchor_outputs`, so they would need to be separate flags (and luckily for us, for now it means we use exactly the flags defined in laolu's dynamic commitment proposal: https://lists.linuxfoundation.org/pipermail/lightning-dev/2020-July/002763.html)

Reply:
> Note that the spec requires that when anchor_output is used, static_remotekey has to be used in conjonction

That's the part I was missing.

So:

| feature bits | result | 
|------------------------|-------------------------|
| `STATIC_REMOTEKEY` | static |
| `ANCHOR_OUTPUTS` | static + anchor |
| `STATIC_REMOTEKEY` \| `ANCHOR_OUTPUTS` | static + anchor |

Reply:
Exactly. Should we explicitly ensure `ANCHOR_OUTPUTS` cannot be used alone with a `require`?

Reply:
> Should we explicitly ensure ANCHOR_OUTPUTS cannot be used alone with a require?

We can't be more restrictive than the spec and it doesn't seem to be very clear on this topic.

Reply:
The spec says that you can't activate feature `option_anchor_outputs` without activating `option_static_remotekey`, which we verify in `validateFeatureGraph`.

We're not yet calling `validateFeatureGraph` at `init` or on `node_announcements`, only when looking at invoice features (because in most cases it's harmless if remote node don't correctly set all feature dependencies). But we also verify this at start-up for our own features, so eclair nodes will not be able to turn on `option_anchor_outputs` without turning on `option_static_remotekey`.

The current code assumes that if `option_anchor_outputs` is turned on, we use a static key. If a remote peer is buggy and doesn't do that, at worst we'll simply close the channel because of a commitment mismatch, so I think it should be fine.

Reply:
> The spec says that you can't activate feature option_anchor_outputs without activating option_static_remotekey, which we verify in validateFeatureGraph.

I missed that, can you point me to it?

> We're not yet calling validateFeatureGraph at init or on node_announcements

Ha, didn't know that, is there a compatibility reason for not doing it today?

Reply:
> I missed that, can you point me to it?

It's in the one-line change to Bolt 9, where it specifies that `option_static_remotekey` is a required dependency for `option_anchor_outputs`.

> Ha, didn't know that, is there a compatibility reason for not doing it today?

Eclair-mobile and Phoenix didn't set all dependencies (because it was added to the spec fairly recently). We have a bunch of backwards-compat' workarounds in the code like https://github.com/ACINQ/eclair/blob/bc81cd5272f93d16cdda20b8f47513e3939733ad/eclair-core/src/main/scala/fr/acinq/eclair/Features.scala#L229

I think we should go towards stricter checks here, but we wanted to give some time for implementations to abide by these transitive feature checks and users to upgrade. I'll verify that latest eclair correctly sets them and verify what eclair-mobile/phoenix do.

Reply:
I know that we currently have the workarounds, but I thought it was precisely because we were enforcing the dependency check as much as possible.

Saw the spec, I'm 100% in favor of adding the dependency check.

Reply:
I'll do tests on our mainnet peers and will open a separate PR to activate that check.

Reply:
Done in https://github.com/ACINQ/eclair/pull/1495